Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1381 @@
+
+Below are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* Ambiguous grammars are reported to the user with parser generators, allowing a greater trust in the parser's output and saving the debugging headache.
+* Bottom-up parsers can handle left-recursion, while top-down can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+
+Cons:
+* Less idiomatic, as you need to write Alex and Happy grammars instead of Haskell code.
+* Less flexible since you can't touch the algorithm on how Happy works. Meanwhile, you can easily create your functions for parser combinators.
+* Dealing with operator precedence and associativity requires more labor. There are standard tools to deal with them, but once you need something that is non-standard, you'll need to write more non-trivial parser rules by hand.
+
+Comparing the implementations of both parsing techniques is beyond the scope of this post. Still, if you are interested, you may read the parser combinators article linked above.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial, but for this, we recommend you read our article, [How to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminals, non-terminals, and production rules
+
+A **terminal** is a lexical item of the language; in our case they will be represented by tokens that were produced by Alex.
+
+A **non-terminal** is an arbitrary identifier that's replaced by other terminals and non-terminals (terminals and non-terminals together are called **symbols**) to produce derivations, i.e., it's a description of how to group symbols together, meaning that a sequence that doesn't match any non-terminal is considered a syntax error.
+In the example below, `myNonTerminal` is non-terminal.
+
+A **production rule** is a rule describing how to "produce" a string that is syntactically valid according to the grammar. It joins terminals together to form a sentence.
+It looks like so in Happy:
+
+```happy
+myNonTerminal
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+n.b.: The ellipses here are not correct Happy grammar, and we use them to represent that more symbols may be present.
+
+`body@#` where `@` is some letter and `#` some number means that the parser must parse each body in succession.
+The pipe (`|`) represents an alternation. The parser may parse any of the given bodies, accepting the one that matches.
+An alternative may be empty, meaning that `myNonTerminal : { action }` is also accepted.
```

Comment:
I don't understand your question, are you asking whether there may multiple empty alternatives? If so, then I guess no, otherwise it would be a reduce/reduce conflict.

Reply:
Well, there isn't a way to phrase it much better 😅 we do call production rules that do not have anything on the rhs "empty productions". Again, we could make the terminology in this section a bit more pedantic and talk about production bodies, then we can have something to the tune of "A body of a production [rule] may be empty".

Reply:
I liked this wording, so I changed it.